src/core_functions/scalar/list/array_slice.cpp

```diff
@@ -1,14 +1,31 @@
 #include "duckdb/core_functions/scalar/list_functions.hpp"
-#include "duckdb/common/pair.hpp"
 #include "duckdb/common/string_util.hpp"
 #include "duckdb/common/types/data_chunk.hpp"
 #include "duckdb/function/scalar/nested_functions.hpp"
 #include "duckdb/function/scalar/string_functions.hpp"
-#include "duckdb/parser/expression/bound_expression.hpp"
 #include "duckdb/planner/expression/bound_function_expression.hpp"
+#include "duckdb/planner/expression/bound_constant_expression.hpp"
 
 namespace duckdb {
 
+static int CalculateSliceLength(int64_t &begin, int64_t &end, int64_t step, bool svalid) {
```

Comment:
To your third point in ClampSlice, the end is set to the `MaxValue(begin, end)`, and since CalculateSliceLength is always called after ClampSlice it should never happen that `end < begin`.